modules/task/execution.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// fnRegister allows looking up a function by name
+type fnRegister struct {
+	fnNameMap map[string]interface{}
+	sync.RWMutex
+}
+
+var fnLookup = fnRegister{fnNameMap: make(map[string]interface{})}
+
+// fnSignature represents a function and its arguments
+type fnSignature struct {
+	FnName string
+	Args   []interface{}
+}
+
+// Execute executes the function
+func (s *fnSignature) Execute() ([]reflect.Value, error) {
+	targetArgs := make([]reflect.Value, 0, len(s.Args))
+	for _, arg := range s.Args {
+		targetArgs = append(targetArgs, reflect.ValueOf(arg))
+	}
+	fnLookup.RLock()
+	defer fnLookup.RUnlock()
+	if fn, ok := fnLookup.fnNameMap[s.FnName]; ok {
+		fnValue := reflect.ValueOf(fn)
+		return fnValue.Call(targetArgs), nil
+	}
+	return nil, fmt.Errorf("function: %q not found. Did you forget to register?", s.FnName)
+}
+
+// Enqueue sends a func before sending to the task queue
+func Enqueue(fn interface{}, args ...interface{}) error {
+	// Validate function
+	fnType := reflect.TypeOf(fn)
+	if err := validateFnFormat(fnType); err != nil {
+		return err
+	}
+	fnName := getFunctionName(fn)
+	// Register
+	if err := register(fn, fnType, fnName); err != nil {
```

Comment:
What is the function registration story?
Can we consume functions that were not registered? E.g. in case of a distributed task system if a function was registered on one machine, but not on another, how would we deal with it? Looks like consume can't be called before all the functions are registered, how can we prevent it?